streamparse/storm/component.py

```diff
@@ -0,0 +1,332 @@
+"""Base primititve classes for working with Storm."""
+from __future__ import absolute_import, print_function, unicode_literals
+
+import io
+import logging
+import os
+import sys
+from collections import deque, namedtuple
+from threading import RLock
+from traceback import format_exc
+
+try:
+    import simplejson as json
+except ImportError:
+    import json
+
+
+# Support for Storm Log levels as per STORM-414
+_STORM_LOG_TRACE = 0
+_STORM_LOG_DEBUG = 1
+_STORM_LOG_INFO = 2
+_STORM_LOG_WARN = 3
+_STORM_LOG_ERROR = 4
+_STORM_LOG_LEVELS = {
+    'trace': _STORM_LOG_TRACE,
+    'debug': _STORM_LOG_DEBUG,
+    'info': _STORM_LOG_INFO,
+    'warn': _STORM_LOG_WARN,
+    'warning': _STORM_LOG_WARN,
+    'error': _STORM_LOG_ERROR,
+}
+_PYTHON_LOG_LEVELS = {
+    'critical': logging.CRITICAL,
+    'error': logging.ERROR,
+    'warning': logging.WARNING,
+    'warn': logging.WARNING,
+    'info': logging.INFO,
+    'debug': logging.DEBUG,
+    'trace': logging.DEBUG
+}
+
+
+log = logging.getLogger(__name__)
+
+
+class StormHandler(logging.Handler):
+    """Handler that will send messages back to Storm."""
+
+    def __init__(self, stream=None):
+        """ Initialize handler """
+        if stream is None:
+            stream = sys.stdout
+        super(StormHandler, self).__init__()
+        self._component = Component(output_stream=stream)
+
+    def emit(self, record):
+        """
+        Emit a record.
+
+        If a formatter is specified, it is used to format the record.
+        If exception information is present, it is formatted using
+        traceback.print_exception and sent to Storm.
+        """
+        try:
+            msg = self.format(record)
+            level = _STORM_LOG_LEVELS.get(record.levelname.lower(),
+                                          _STORM_LOG_INFO)
+            self._component.send_message({'command': 'log', 'msg': str(msg),
+                                          'level': level})
+        except Exception:
+            self.handleError(record)
+
+
+class LogStream(object):
+    """Object that implements enough of the Python stream API to be used as
+    sys.stdout. Messages are written to the Python logger.
+    """
+    def __init__(self, logger):
+        self.logger = logger
+
+    def write(self, message):
+        if message.strip() == "":
+            return  # skip blank lines
+
+        try:
+            self.logger.info(message)
+        except:
+            # There's been an issue somewhere in the logging sub-system
+            # so we'll put stderr and stdout back to their originals and
+            # raise the exception which will cause Storm to choke
+            sys.stdout = sys.__stdout__
+            raise
+
+    def flush(self):
+        """No-op method to prevent crashes when someone does
+        sys.stdout.flush.
+        """
+        pass
+
+
+Tuple = namedtuple('Tuple', 'id component stream task values')
```

Comment:
When we were explicitly defining `Tuple` as just a class that inherited from `object` (and set `__slots__`), we didn't set things up such that `__eq__` worked, so this seemed a simpler approach to getting what we wanted (since `namedtuple` classes always have `__slots__` set).